README.md
![repo-banner](./assets/wobble-logo.png)

[![Current version:](https://img.shields.io/badge/v1.5.1:-222222.svg?logo=npm)](https://www.npmjs.com/package/wobble/v/1.5.1)
[![Test status](https://img.shields.io/circleci/project/github/skevy/wobble/stable.svg?logo=circleci&label=Tests)](https://circleci.com/gh/skevy/wobble/133)
[![Code coverage](https://img.shields.io/codecov/c/github/skevy/wobble/stable.svg?logo=codecov&logoColor=white&label=Coverage)](https://codecov.io/gh/skevy/wobble/tree/81d48aa583e633fef9da5d0050002b7438a1eb4d/src)<br />
[![HEAD:](https://img.shields.io/badge/HEAD:-222222.svg?logo=github&logoColor=white)](https://github.com/skevy/wobble)
[![Test status](https://img.shields.io/circleci/project/github/skevy/wobble/develop.svg?logo=circleci&label=Tests)](https://circleci.com/gh/skevy/wobble/tree/develop)
[![Code coverage](https://img.shields.io/codecov/c/github/skevy/wobble/develop.svg?logo=codecov&logoColor=white&label=Coverage)](https://codecov.io/gh/skevy/wobble/branch/develop)

A tiny (~1.7 KB gzipped) spring physics micro-library that models a [damped harmonic oscillator](https://en.wikipedia.org/wiki/Harmonic_oscillator#Damped_harmonic_oscillator).

# Why Would I Use This?

Use _wobble_ if you need a very small and accurate damped harmonic spring simulation in your animation library or application. _wobble_ intentionally **only** provides a way to animate a scalar value according to equations governing damped harmonic motion. That's all this library will ever do -- any other functionality (integration with [insert ui library here], multi-dimensional springs, a nice API around chaining springs together, etc.) is left to the reader to implement.

# Background

Using spring physics in UI design is a common way to express natural motion, and there are several ways to model the physics behind springs.

There are two main ways that spring physics is typically implemented: numerical integration (using something like the [Runge-Kutta 4th order](https://en.wikipedia.org/wiki/Runge%E2%80%93Kutta_methods) numerical integration method) or by using a closed-form (exact) solution. Numerical integration is an approximation of the exact solution, and is generally easier to derive. The numerical integration technique can be applied to basically any ordinary differential equation. Though there are several different numerical integration methods, it's common to leverage RK4 when accuracy of the approximation is required, even though it's slightly slower. RK4 is commonly used in other animation libraries, such as [Rebound](https://github.com/facebook/rebound) and [Pop](https://github.com/facebook/pop/).

The original goal of the algorithm used in _wobble_ was to replicate `CASpringAnimation` from Apple's Core Animation library (used to power animations on macOS and iOS) in order to mimic iOS animations in React Native. After doing a [little spelunking](https://github.com/facebook/react-native/pull/15322) inside `QuartzCore.framework`, it became clear that Apple was using the closed-form solution for damped harmonic oscillation to power `CASpringAnimation`. _wobble_ leverages the same equations as `CASpringAnimation` in order to be able to match Apple animations precisely.

The closed-form solution lets us calculate position (_x_) and velocity (_v_) from time _t_, and thus it turns out that using the closed-form solution provides a couple advantages over RK4:

- Easier to generate keyframes and build continuous/interruptible gestures and animations, due to the fact that _x_ and _v_ are pure functions of _t_.
- Less code.
- It's faster.

### Math!

The ODE for damped harmonic motion is:

![DHO ODE](./assets/ode.gif)

Solving this ODE yields:

![solution](./assets/general-solution.gif)

And from this general solution, we're able to easily derive the solutions for under-damped, critically-damped, and over-damped damped harmonic oscillation.

The full proof can be found [in this PDF](http://planetmath.org/sites/default/files/texpdf/39745.pdf) from [planetmath.org](http://planetmath.org).

# Demos

Wobble demos are located here: [https://wobble-demos.now.sh/](https://wobble-demos.now.sh/). Send PRs to add more!

# Getting Started

```bash
yarn add wobble
# or
npm install --save wobble
```

# Usage

```jsx
import { Spring } from 'wobble';

// Create a new spring
const spring = new Spring({
  toValue: 100,
  stiffness: 1000,
  damping: 500,
  mass: 3,
});

// Set listeners for spring events, start the spring.
spring
  .onStart(() => {
    console.log('Spring started!');
  })
  .onUpdate((s) => {
    console.log(`Spring's current value: ` + s.currentValue);
    console.log(`Spring's current velocity: ` + s.currentVelocity);
  })
  .onStop(() => {
    console.log('Spring is at rest!');
  })
  .start();
```

# API

#### `new Spring(config: SpringConfig)`

Initialize a new spring with a given spring configuration.

## Configuration

#### `fromValue: number`

Starting value of the animation. Defaults to `0`.

#### `toValue: number`

Ending value of the animation. Defaults to `1`.

#### `stiffness: number`

The spring stiffness coefficient. Defaults to `100`.

#### `damping: number`

Defines how the spring’s motion should be damped due to the forces of friction. Defaults to `10`.

#### `mass: number`

The mass of the object attached to the end of the spring. Defaults to `1`.

#### `initialVelocity: number`

The initial velocity (in units/ms) of the object attached to the spring. Defaults to `0`.

#### `allowsOverdamping: boolean`

Whether or not the spring allows "overdamping" (a damping ratio > 1). Defaults to `false`.

#### `overshootClamping: boolean`

False when overshooting is allowed, true when it is not. Defaults to `false`.

#### `restVelocityThreshold: number`

When spring's velocity is below `restVelocityThreshold`, it is at rest. Defaults to `.001`.

#### `restDisplacementThreshold: number`

When the spring's displacement (current value) is below `restDisplacementThreshold`, it is at rest. Defaults to `.001`.

## Methods

#### `start(): Spring`

If `fromValue` differs from `toValue`, or `initialVelocity` is non-zero, start the simulation and call the `onStart` listeners.

#### `stop(): Spring`

If a simulation is in progress, stop it and call the `onStop` listeners.

#### `updateConfig(updatedConfig: PartialSpringConfig): Spring`

Updates the spring config with the given values. Values not explicitly supplied will be reused from the existing config.

#### `onStart(listener: SpringListenerFn): Spring`

The provided callback will be invoked when the simulation begins.

#### `onUpdate(listener: SpringListenerFn): Spring`

The provided callback will be invoked on each frame while the simulation is running.

#### `onStop(listener: SpringListenerFn): Spring`

The provided callback will be invoked when the simulation ends.

#### `removeListener(listenerFn: SpringListenerFn): Spring`

Remove a single listener from this spring.

#### `removeAllListeners(): Spring`

Removes all listeners from this spring.

## Properties

#### `currentValue: number`

The spring's current displacement.

#### `currentVelocity: number`

The spring's current velocity in units / ms.

#### `isAtRest: boolean`

If the spring has reached its `toValue`, or if its velocity is below the `restVelocityThreshold`, it is considered at rest. If `stop()` is called during a simulation, both `isAnimating` and `isAtRest` will be false.

#### `isAnimating: boolean`

Whether or not the spring is currently emitting values. Note: this is distinct from whether or not it is at rest. See also `isAtRest`.

# Credits

Brenton Simpson ([@appsforartists](https://twitter.com/appsforartists)) - For his assistance in creating and testing this library.

Devine Lu Linvega ([@neauoire](https://twitter.com/neauoire)) - The awesome logo!


CHANGELOG.md
# CHANGELOG

### v1.5.1

- Distribute unminified bundles by default - minified versions are available at `/dist/wobble.${MODULE_FORMAT}.min.js` ([af82dda](https://github.com/skevy/wobble/commit/af82dda29eae11741efd380fcde5a1e7542f759d))
- Add `typescript:main` field to `package.json`, which points to the raw TypeScript source ([0198d63](https://github.com/skevy/wobble/commit/0198d630a4d426c9fc9bbdb1e7e9e74c28873c31))

### v1.5.0

- Fixup `yarn watch` to support es module compilation ([f7dd689](https://github.com/skevy/wobble/commit/f7dd68951e0f2cd8a083706c9727b1906d3b6e03))
- Add module entry point for .es.js dist ([0b929acf](https://github.com/skevy/wobble/commit/0b929acf6ba3a443b974a1c350213a3d64ff45cc))

### v1.4.1

- Remove Babel ([cfdc3d85](https://github.com/skevy/wobble/commit/cfdc3d85ce01687b77c44cc68d11cabc85ffe0a3))
- Fixup typescript interface names and member access ([2112b118](https://github.com/skevy/wobble/commit/2112b118fbb491113ec0432d9080b7f7acfc2bd5))

### v1.4.0

- Remove Flow, rewrite in TypeScript ([a85aa77](https://github.com/skevy/wobble/commit/a85aa771843ec9a9ddf7d3d187dc4c2e5ff5a1b6))

### v1.3.1

- Fix types for chainable methods ([bd7b496](https://github.com/skevy/wobble/commit/bd7b49622805b9bc218c87017ea92d3cb35d623f))
- Use lolex in tests to properly fake timers ([3cb1e92](https://github.com/skevy/wobble/commit/3cb1e92729eb7948776f4548f9e86a4bad4c3077))

### v1.3.0

- Update TypeScript definition with chainable API ([0c9bb96](https://github.com/skevy/wobble/commit/0c9bb96eb66c716d4d3b057bb366d593ea9bc79e))
- Add README ([4656acf](https://github.com/skevy/wobble/commit/4656acf67f1a892ac81a1684d6516384eb900cd3
))
- Deploy demos to wobble-demos.now.sh ([4172c9c](https://github.com/skevy/wobble/commit/4172c9c13aa6877a7f3951c201ca33a4e3e37856
))
- Set _currentValue + _currentVelocity in constructor ([6112980](https://github.com/skevy/wobble/commit/61129802df8b2536264ceff46823e5e3911e9eef))
- Add "chat heads" demo ([5189aea](https://github.com/skevy/wobble/commit/5189aea291131816fd80dc834fac2885a72e58b1))
- Use `Date.now()` instead of `performance.now()` for React Native compatibility ([34d9a0f](https://github.com/skevy/wobble/commit/34d9a0fe6ef3ac24630c8524a0ac1229ed09fc08))
- Make `this._currentTime` always equal to "now", remove "first step" calculation ([4589723](https://github.com/skevy/wobble/commit/45897231dd8683f4e820d2030956fe34cb4866ee))
- Make all APIs chainable ([b89559f](https://github.com/skevy/wobble/commit/b89559fb6e83314098a673461d2225369f170d81))
- Set default `toValue` to 1. ([c0fb41a](https://github.com/skevy/wobble/commit/c0fb41a3d17f30fb1c5298cc3c2aa20af7fe55a8))
- Move @flow annotations + fix comments ([058c75d](https://github.com/skevy/wobble/commit/058c75d934a36c0e2a8f5dd5f34d5854ce664124))
- Remove "browser" field from package.json ([261cca3](https://github.com/skevy/wobble/commit/261cca3237837cf2650a9441fe1406bc41a37b73))

### v1.2.0

- Consistently use the term "listener" in tests ([1328f38](https://github.com/skevy/wobble/commit/1328f3878b7eea60611f4d722fa13d32e854826f))
- Ensure listeners are only notified once per frame ([cf92f76](https://github.com/skevy/wobble/commit/cf92f7670aae10dc776af1a624a5d2509d0f1b70))
- Made _evaluateSpring take an absolute timestamp ([0696712](https://github.com/skevy/wobble/commit/0696712d1352263d6dc09bc61e03be06be4f98ed))

### v1.1.0

- Made onStart synchronous ([aaf21a9](https://github.com/skevy/wobble/commit/aaf21a97435864b93122ed6480765a654713f888))
- Update _currentValue and _currentVelocity before updating fromValue ([9abf13f](https://github.com/skevy/wobble/commit/9abf13f8e1a107619a31c304750cd0363ad976ed))

### v1.0.0

🎉🎉🎉🎉🎉🎉 Initial Release! 🎉🎉🎉🎉🎉🎉


src/__tests__/Springs-test.js
/** @license
 *  Copyright 2017 Adam Miskiewicz
 *
 *  Use of this source code is governed by a MIT-style license that can be found
 *  in the LICENSE file or at https://opensource.org/licenses/MIT.
 *  
 *  @flow
 */

import lolex from "lolex";
import { Spring } from "../";

let clock;

describe("Spring", () => {
  it("animates correctly with base values", () => {
    const clock = lolex.install();

    const spring = new Spring({
      fromValue: 0,
      toValue: 1
    });

    const { values, velocities } = _measureSpringDynamics(clock, spring, 80);
    expect(values).toMatchSnapshot();
    expect(velocities).toMatchSnapshot();

    clock.uninstall();
  });

  it("animates correctly with non-zero initialVelocity", () => {
    const clock = lolex.install();

    const spring = new Spring({
      fromValue: 10,
      toValue: 100,
      stiffness: 230,
      damping: 22,
      mass: 1,
      initialVelocity: 2
    });

    const { values, velocities } = _measureSpringDynamics(clock, spring, 80);

    expect(velocities[0]).toBe(2);
    expect(velocities[values.length - 1]).toBe(0);

    expect(values).toMatchSnapshot();
    expect(velocities).toMatchSnapshot();

    clock.uninstall();
  });

  it("animates correctly with non-zero fromValue", () => {
    const clock = lolex.install();

    const spring = new Spring({
      fromValue: 25,
      toValue: 50,
      stiffness: 230,
      damping: 22,
      mass: 1
    });

    const { values, velocities } = _measureSpringDynamics(clock, spring, 60);

    expect(values[0]).toBe(25);
    expect(values[values.length - 1]).toBe(50);

    expect(values).toMatchSnapshot();
    expect(velocities).toMatchSnapshot();

    clock.uninstall();
  });

  it("animates correctly with fromValue that's larger than toValue", () => {
    const clock = lolex.install();

    const spring = new Spring({
      fromValue: 200,
      toValue: 11,
      stiffness: 230,
      damping: 22,
      mass: 1
    });

    const { values, velocities } = _measureSpringDynamics(clock, spring, 60);

    expect(values[0]).toBe(200);
    expect(values[values.length - 1]).toBe(11);

    expect(values).toMatchSnapshot();
    expect(velocities).toMatchSnapshot();

    clock.uninstall();
  });

  it("keeps proper dynamics when the config is updated during the animation", () => {
    const clock = lolex.install({
      global: global
    });

    const spring = new Spring({
      fromValue: 25,
      toValue: 131,
      stiffness: 230,
      damping: 22,
      mass: 1
    });

    const {
      values: expectedValues,
      velocities: expectedVelocities
    } = _measureSpringDynamics(clock, spring, 60);

    // Just run spring for 10 frames...we don't care about its
    // values/velocities -- other tests cover this
    _measureSpringDynamics(clock, spring, 10);

    spring.updateConfig({}); // don't change anything

    const {
      values: valuesAfterUpdate,
      velocities: velocitiesAfterUpdate
    } = _measureSpringDynamics(clock, spring, 60);

    // Get the expected values/velocities after 10 frames
    const expectedValuesAfterUpdate = expectedValues.slice(10);
    const expectedVelocitiesAfterUpdate = expectedVelocities.slice(10);

    valuesAfterUpdate.forEach((val, i) => {
      expect(val).toBeCloseTo(expectedValuesAfterUpdate[i], 0.00000001);
    });
    velocitiesAfterUpdate.forEach((vel, i) => {
      expect(vel).toBeCloseTo(expectedVelocitiesAfterUpdate[i], 0.00000001);
    });

    clock.uninstall();
  });

  it("respects toValue being updated mid-animation", () => {
    const clock = lolex.install();

    const spring = new Spring({
      fromValue: 25,
      toValue: 131,
      stiffness: 230,
      damping: 22,
      mass: 1
    });

    const {
      values: expectedValues,
      velocities: expectedVelocities
    } = _measureSpringDynamics(clock, spring, 60);

    // Just run spring for 5 frames
    _measureSpringDynamics(clock, spring, 5);

    spring.updateConfig({
      toValue: 200,
      restDisplacementThreshold: 0.01
    }); // Extend the spring

    // Run the simulation till spring rests
    const {
      values: valuesAfterUpdate,
      velocities: velocitiesAfterUpdate
    } = _measureSpringDynamics(clock, spring, 60);

    // Expect spring to be at rest with new values
    expect(valuesAfterUpdate[valuesAfterUpdate.length - 1]).toBe(200);
    expect(velocitiesAfterUpdate[velocitiesAfterUpdate.length - 1]).toBe(0);

    clock.uninstall();
  });

  it("respects initialVelocity being updated mid-animation", () => {
    const clock = lolex.install();

    const spring = new Spring({
      fromValue: 25,
      toValue: 131,
      stiffness: 230,
      damping: 22,
      mass: 1
    });

    const {
      values: expectedValues,
      velocities: expectedVelocities
    } = _measureSpringDynamics(clock, spring, 60);

    // Just run spring for 10 frames
    _measureSpringDynamics(clock, spring, 10);

    spring.updateConfig({
      initialVelocity: 200
    });

    // Run the simulation till spring rests
    const {
      values: valuesAfterUpdate,
      velocities: velocitiesAfterUpdate
    } = _measureSpringDynamics(clock, spring, 60);

    // Expect spring to be at rest with new values
    expect(valuesAfterUpdate[0]).toBe(expectedValues[10]); // first frame should be the same position as ending frame
    expect(valuesAfterUpdate[1]).toBeGreaterThan(2000); // should go to a really high number
    expect(velocitiesAfterUpdate[0]).toBe(200);
    expect(velocitiesAfterUpdate[8]).toBeLessThan(0); // given the spring dynamics, v should go negative pretty quickly

    clock.uninstall();
  });

  it("removes existing listener", () => {
    const spring = new Spring();
    const onUpdateListener = () => {};
    spring.onUpdate(onUpdateListener);
    expect(spring._listeners[0].onUpdate).toBe(onUpdateListener);
    spring.removeListener(onUpdateListener);
    expect(spring._listeners).toEqual([]);
  });

  it("clears all listeners", () => {
    const spring = new Spring();
    const onUpdateListener = () => {};
    spring.onUpdate(onUpdateListener);
    expect(spring._listeners[0].onUpdate).toBe(onUpdateListener);
    spring.removeAllListeners();
    expect(spring._listeners).toEqual([]);
  });

  it("synchronously notifies listeners that the spring has started when .start() is called", () => {
    const onStartListener = jest.fn();

    const spring = new Spring({
      toValue: 1
    });

    spring.onStart(onStartListener);

    spring.start();
    expect(onStartListener).toHaveBeenCalledWith(spring);
  });

  it("synchronously notifies listeners that the spring has stopped when .stop() is called mid-animation", () => {
    const clock = lolex.install();
    const onStopListener = jest.fn();

    const spring = new Spring({
      toValue: 1
    });

    spring
      .onStop(s => {
        onStopListener(s);
        expect(s.currentVelocity).toBeCloseTo(0.00165009062976268, 0.001);
      })
      .start();

    clock.tick(1000 / 60 * 10);

    spring.stop();
    expect(onStopListener).toHaveBeenCalledWith(spring);
    expect(spring.isAtRest).toBeFalsy();

    clock.tick(1000 / 60 * 1);

    expect(onStopListener).toHaveBeenCalledTimes(1);

    clock.uninstall();
  });

  it("should only call onUpdate once per frame", () => {
    const clock = lolex.install();
    const onUpdateListener = jest.fn();
    const onStopListener = jest.fn();

    const spring = new Spring({ fromValue: 0, toValue: 1 });
    spring.onUpdate(onUpdateListener);
    spring.onStop(onStopListener);

    spring.start();
    spring.updateConfig({ toValue: 0 });
    spring.updateConfig({ fromValue: 1 });

    expect(onUpdateListener).not.toHaveBeenCalled();

    clock.tick(1000 / 60);

    expect(onUpdateListener).toHaveBeenCalledTimes(1);
    expect(onStopListener).not.toHaveBeenCalled();

    clock.uninstall();
  });

  it("should call onStop on the next frame if fromValue and toValue are updated to match during the animation", () => {
    const clock = lolex.install();
    const onStopListener = jest.fn();

    const spring = new Spring({ fromValue: 0, toValue: 1 });
    spring.onStop(onStopListener);

    spring.start();

    clock.tick(1000 / 60);
    spring.updateConfig({ fromValue: 1, initialVelocity: 0 });
    expect(onStopListener).not.toHaveBeenCalled();

    clock.tick(1000 / 60);
    expect(onStopListener).toHaveBeenCalled();

    clock.uninstall();
  });

  function _measureSpringDynamics(
    clock: any,
    spring: Spring,
    numFrames: number
  ) {
    const values = [spring.currentValue];
    const velocities = [spring.currentVelocity];

    spring
      .onUpdate(spring => {
        values.push(spring.currentValue);
        velocities.push(spring.currentVelocity);
      })
      .start();

    clock.tick(Math.round(1000 / 60.0 * numFrames / 16) * 16); // round to the nearest frame

    spring.removeAllListeners();

    return { values, velocities };
  }
});


jest/rAF.js
// http://paulirish.com/2011/requestanimationframe-for-smart-animating/
// http://my.opera.com/emoller/blog/2011/12/20/requestanimationframe-for-smart-er-animating

// requestAnimationFrame polyfill by Erik Möller. fixes from Paul Irish and Tino Zijdel

// MIT license

(function() {
  var lastTime = 0;
  var vendors = ["ms", "moz", "webkit", "o"];
  for (var x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
    window.requestAnimationFrame = window[vendors[x] + "RequestAnimationFrame"];
    window.cancelAnimationFrame =
      window[vendors[x] + "CancelAnimationFrame"] ||
      window[vendors[x] + "CancelRequestAnimationFrame"];
  }

  if (!window.requestAnimationFrame)
    window.requestAnimationFrame = function(callback, element) {
      var currTime = new Date().getTime();
      var timeToCall = Math.max(0, 16 - (currTime - lastTime));
      var id = window.setTimeout(function() {
        callback(currTime + timeToCall);
      }, 1000 / 60);
      lastTime = currTime + timeToCall;
      return id;
    };

  if (!window.cancelAnimationFrame)
    window.cancelAnimationFrame = function(id) {
      clearTimeout(id);
    };

  if (!window.performance)
    window.performance = {
      now() {
        // Using rAF polyfill's `lastTime` to ensure that timestamp passed into
        // rAF's callback and `performance.now()` are on the same timescale.
        return lastTime;
      }
    };
})();


src/index.ts
/**
 *  @license
 *  Copyright 2017 Adam Miskiewicz
 *
 *  Use of this source code is governed by a MIT-style license that can be found
 *  in the LICENSE file or at https://opensource.org/licenses/MIT.
 */

import { invariant, withDefault } from "./utils";

export interface SpringConfig {
  fromValue: number; // Starting value of the animation.
  toValue: number; // Ending value of the animation.
  stiffness: number; // The spring stiffness coefficient.
  damping: number; // Defines how the spring’s motion should be damped due to the forces of friction.
  mass: number; // The mass of the object attached to the end of the spring.
  initialVelocity: number; // The initial velocity (in units/ms) of the object attached to the spring.
  allowsOverdamping: boolean; // Whether or not the spring allows "overdamping" (a damping ratio > 1). Defaults to false.
  overshootClamping: boolean; // False when overshooting is allowed, true when it is not. Defaults to false.
  restVelocityThreshold: number; // When spring's velocity is below `restVelocityThreshold`, it is at rest. Defaults to .001.
  restDisplacementThreshold: number; // When the spring's displacement (current value) is below `restDisplacementThreshold`, it is at rest. Defaults to .001.
  [index: string]: any;
}

export type PartialSpringConfig = Partial<SpringConfig>;

export type SpringListenerFn = (spring: Spring) => void;
export interface SpringListener {
  onUpdate?: SpringListenerFn;
  onStart?: SpringListenerFn;
  onStop?: SpringListenerFn;
}

/**
 * Implements a spring physics simulation based on the equations behind
 * damped harmonic oscillators (https://en.wikipedia.org/wiki/Harmonic_oscillator#Damped_harmonic_oscillator).
 */
export class Spring {
  static MAX_DELTA_TIME_MS = 1 / 60 * 1000 * 4; // advance 4 frames at max

  _listeners: SpringListener[] = [];

  private _config: SpringConfig;
  private _currentAnimationStep: number = 0; // current requestAnimationFrame

  private _currentTime: number = 0; // Current timestamp of animation in ms (real time)
  private _springTime: number = 0; // Current time along the spring curve in ms (zero-based)

  private _currentValue: number = 0; // the current value of the spring
  private _currentVelocity: number = 0; // the current velocity of the spring
  private _isAnimating: boolean = false;

  private _oscillationVelocityPairs = [];

  constructor(config: PartialSpringConfig = {}) {
    this._config = {
      fromValue: withDefault(config.fromValue, 0),
      toValue: withDefault(config.toValue, 1),
      stiffness: withDefault(config.stiffness, 100),
      damping: withDefault(config.damping, 10),
      mass: withDefault(config.mass, 1),
      initialVelocity: withDefault(config.initialVelocity, 0),
      overshootClamping: withDefault(config.overshootClamping, false),
      allowsOverdamping: withDefault(config.allowsOverdamping, false),
      restVelocityThreshold: withDefault(config.restVelocityThreshold, 0.001),
      restDisplacementThreshold: withDefault(
        config.restDisplacementThreshold,
        0.001
      )
    };
    this._currentValue = this._config.fromValue;
    this._currentVelocity = this._config.initialVelocity;
  }

  /**
   * If `fromValue` differs from `toValue`, or `initialVelocity` is non-zero,
   * start the simulation and call the `onStart` listeners.
   */
  start(): this {
    const { fromValue, toValue, initialVelocity } = this._config;

    if (fromValue !== toValue || initialVelocity !== 0) {
      this._reset();
      this._isAnimating = true;

      if (!this._currentAnimationStep) {
        this._notifyListeners("onStart");
        this._currentAnimationStep = requestAnimationFrame((t: number) => {
          this._step(Date.now());
        });
      }
    }

    return this;
  }

  /**
   * If a simulation is in progress, stop it and call the `onStop` listeners.
   */
  stop(): this {
    if (!this._isAnimating) {
      return this;
    }

    this._isAnimating = false;
    this._notifyListeners("onStop");

    if (this._currentAnimationStep) {
      cancelAnimationFrame(this._currentAnimationStep);
      this._currentAnimationStep = 0;
    }

    return this;
  }

  /**
   * The spring's current position.
   */
  get currentValue(): number {
    return this._currentValue;
  }

  /**
   * The spring's current velocity in units / ms.
   */
  get currentVelocity(): number {
    return this._currentVelocity; // give velocity in units/ms;
  }

  /**
   * If the spring has reached its `toValue`, or if its velocity is below the
   * `restVelocityThreshold`, it is considered at rest. If `stop()` is called
   * during a simulation, both `isAnimating` and `isAtRest` will be false.
   */
  get isAtRest(): boolean {
    return this._isSpringAtRest();
  }

  /**
   * Whether or not the spring is currently emitting values.
   *
   * Note: this is distinct from whether or not it is at rest.
   * See also `isAtRest`.
   */
  get isAnimating(): boolean {
    return this._isAnimating;
  }

  /**
   * Updates the spring config with the given values.  Values not explicitly
   * supplied will be reused from the existing config.
   */
  updateConfig(updatedConfig: PartialSpringConfig): this {
    // When we update the spring config, we reset the simulation to ensure the
    // spring always moves the full distance between `fromValue` and `toValue`.
    // To ensure that the simulation behaves correctly if those values aren't
    // being changed in `updatedConfig`, we run the simulation with `_step()`
    // and default `fromValue` and `initialVelocity` to their current values.

    this._advanceSpringToTime(Date.now());

    this._config.fromValue = this._currentValue;
    this._config.initialVelocity = this._currentVelocity;

    for (const key in updatedConfig) {
      if (this._config.hasOwnProperty(key)) {
        this._config[key] = updatedConfig[key as keyof SpringConfig];
      }
    }

    this._reset();

    return this;
  }

  /**
   * The provided callback will be invoked when the simulation begins.
   */
  onStart(listener: SpringListenerFn): this {
    this._listeners.push({ onStart: listener });
    return this;
  }

  /**
   * The provided callback will be invoked on each frame while the simulation is
   * running.
   */
  onUpdate(listener: SpringListenerFn): this {
    this._listeners.push({ onUpdate: listener });
    return this;
  }

  /**
   * The provided callback will be invoked when the simulation ends.
   */
  onStop(listener: SpringListenerFn): this {
    this._listeners.push({ onStop: listener });
    return this;
  }

  /**
   * Remove a single listener from this spring.
   */
  removeListener(listenerFn: SpringListenerFn): this {
    this._listeners = this._listeners.reduce(
      (result, listener) => {
        const foundListenerFn =
          Object.values(listener).indexOf(listenerFn) !== -1;
        if (!foundListenerFn) {
          result.push(listener);
        }
        return result;
      },
      [] as SpringListener[]
    );
    return this;
  }

  /**
   * Removes all listeners from this spring.
   */
  removeAllListeners(): this {
    this._listeners = [];
    return this;
  }

  private _reset() {
    this._currentTime = Date.now();
    this._springTime = 0.0;
    this._currentValue = this._config.fromValue;
    this._currentVelocity = this._config.initialVelocity;
  }

  private _notifyListeners(eventName: keyof SpringListener) {
    this._listeners.forEach((listener: Partial<SpringListener>) => {
      const maybeListenerFn = listener[eventName];
      if (typeof maybeListenerFn === "function") {
        maybeListenerFn(this);
      }
    });
  }

  /**
   * `_step` is the main loop.  While the animation is running, it updates the
   * current state once per frame, and schedules the next frame if the spring is
   * not yet at rest.
   */
  private _step(timestamp: number) {
    this._advanceSpringToTime(timestamp, true);

    // check `_isAnimating`, in case `stop()` got called during
    // `_advanceSpringToTime()`
    if (this._isAnimating) {
      this._currentAnimationStep = requestAnimationFrame((t: number) =>
        this._step(Date.now())
      );
    }
  }

  private _advanceSpringToTime(
    timestamp: number,
    shouldNotifyListeners: boolean = false
  ) {
    // `_advanceSpringToTime` updates `_currentTime` and triggers the listeners.
    // Because of these side effects, it's only safe to call when an animation
    // is already in-progress.
    if (!this._isAnimating) {
      return;
    }

    let deltaTime = timestamp - this._currentTime;

    // If for some reason we lost a lot of frames (e.g. process large payload or
    // stopped in the debugger), we only advance by 4 frames worth of
    // computation and will continue on the next frame. It's better to have it
    // running at slower speed than jumping to the end.
    if (deltaTime > Spring.MAX_DELTA_TIME_MS) {
      deltaTime = Spring.MAX_DELTA_TIME_MS;
    }
    this._springTime += deltaTime;

    const c = this._config.damping;
    const m = this._config.mass;
    const k = this._config.stiffness;
    const fromValue = this._config.fromValue;
    const toValue = this._config.toValue;
    const v0 = -this._config.initialVelocity;

    invariant(m > 0, "Mass value must be greater than 0");
    invariant(k > 0, "Stiffness value must be greater than 0");
    invariant(c > 0, "Damping value must be greater than 0");

    let zeta = c / (2 * Math.sqrt(k * m)); // damping ratio (dimensionless)
    const omega0 = Math.sqrt(k / m) / 1000; // undamped angular frequency of the oscillator (rad/ms)
    const omega1 = omega0 * Math.sqrt(1.0 - zeta * zeta); // exponential decay
    const omega2 = omega0 * Math.sqrt(zeta * zeta - 1.0); // frequency of damped oscillation
    const x0 = toValue - fromValue; // initial displacement of the spring at t = 0

    if (zeta > 1 && !this._config.allowsOverdamping) {
      zeta = 1;
    }

    let oscillation = 0.0;
    let velocity = 0.0;
    const t = this._springTime;
    if (zeta < 1) {
      // Under damped
      const envelope = Math.exp(-zeta * omega0 * t);
      oscillation =
        toValue -
        envelope *
          ((v0 + zeta * omega0 * x0) / omega1 * Math.sin(omega1 * t) +
            x0 * Math.cos(omega1 * t));
      // This looks crazy -- it's actually just the derivative of the
      // oscillation function
      velocity =
        zeta *
          omega0 *
          envelope *
          (Math.sin(omega1 * t) * (v0 + zeta * omega0 * x0) / omega1 +
            x0 * Math.cos(omega1 * t)) -
        envelope *
          (Math.cos(omega1 * t) * (v0 + zeta * omega0 * x0) -
            omega1 * x0 * Math.sin(omega1 * t));
    } else if (zeta === 1) {
      // Critically damped
      const envelope = Math.exp(-omega0 * t);
      oscillation = toValue - envelope * (x0 + (v0 + omega0 * x0) * t);
      velocity =
        envelope * (v0 * (t * omega0 - 1) + t * x0 * (omega0 * omega0));
    } else {
      // Overdamped
      const envelope = Math.exp(-zeta * omega0 * t);
      oscillation =
        toValue -
        envelope *
          ((v0 + zeta * omega0 * x0) * Math.sinh(omega2 * t) +
            omega2 * x0 * Math.cosh(omega2 * t)) /
          omega2;
      velocity =
        envelope *
          zeta *
          omega0 *
          (Math.sinh(omega2 * t) * (v0 + zeta * omega0 * x0) +
            x0 * omega2 * Math.cosh(omega2 * t)) /
          omega2 -
        envelope *
          (omega2 * Math.cosh(omega2 * t) * (v0 + zeta * omega0 * x0) +
            omega2 * omega2 * x0 * Math.sinh(omega2 * t)) /
          omega2;
    }

    this._currentTime = timestamp;
    this._currentValue = oscillation;
    this._currentVelocity = velocity;

    if (!shouldNotifyListeners) {
      return;
    }

    this._notifyListeners("onUpdate");
    if (!this._isAnimating) {
      // a listener might have stopped us in _onUpdate
      return;
    }

    // If the Spring is overshooting (when overshoot clamping is on), or if the
    // spring is at rest (based on the thresholds set in the config), stop the
    // animation.
    if (this._isSpringOvershooting() || this._isSpringAtRest()) {
      if (k !== 0) {
        // Ensure that we end up with a round value
        this._currentValue = toValue;
        this._currentVelocity = 0;
        this._notifyListeners("onUpdate");
      }

      this.stop();
      return;
    }
  }

  private _isSpringOvershooting() {
    const { stiffness, fromValue, toValue, overshootClamping } = this._config;
    let isOvershooting = false;
    if (overshootClamping && stiffness !== 0) {
      if (fromValue < toValue) {
        isOvershooting = this._currentValue > toValue;
      } else {
        isOvershooting = this._currentValue < toValue;
      }
    }
    return isOvershooting;
  }

  private _isSpringAtRest() {
    const {
      stiffness,
      toValue,
      restDisplacementThreshold,
      restVelocityThreshold
    } = this._config;

    const isNoVelocity =
      Math.abs(this._currentVelocity) <= restVelocityThreshold;
    const isNoDisplacement =
      stiffness !== 0 &&
      Math.abs(toValue - this._currentValue) <= restDisplacementThreshold;
    return isNoDisplacement && isNoVelocity;
  }
}
